docs/pages/architecture/tls-routing.mdx

```diff
@@ -0,0 +1,154 @@
+(e.g. SSH).
+
+To implement TLS routing, Teleport uses SNI ([Server Name Indication](https://en.wikipedia.org/wiki/Server_Name_Indication))
+and ALPN ([Application-Level Protocol Negotiation](https://en.wikipedia.org/wiki/Application-Layer_Protocol_Negotiation))
+TLS extensions.
+
+## How it works
+
+Starting from version `8.0` Teleport proxy listens for all client connections
+on its `web_listen_addr` by default:
+
+```yaml
+proxy_service:
+  web_listen_addr: "0.0.0.0:443"
+```
+
+All Teleport clients including SSH, web browser, kubectl, database and reverse
+tunnel clients establish a TLS tunnel to the proxy's web port and indicate the
+protocol they're requesting using SNI and ALPN TLS extensions.
+
+Upon accepting a new connection, the proxy inspects the SNI/ALPN value in the
+TLS handshake and forwards the connection to appropriate backend service.
+
+### Local proxy
+
+Many external clients, for example database clients like `psql` or `mysql`, do
+not speak plain TLS and instead implement TLS handshake as a part of their
+protocol-specific connection negotiation phase (aka [STARTTLS](https://en.wikipedia.org/wiki/Opportunistic_TLS)).
+
+To support these clients, as well as clients that do speak TLS but don't allow
+setting custom ALPN values, Teleport's `tsh` client includes ability to start
+a local TLS routing aware proxy.
+
+Such clients connect to the local proxy instead of Teleport proxy directly. The
+local proxy establishes a TLS connection to the Teleport proxy with the proper
+SNI/ALPN values set and tunnels the client's connection over it.
+
+Setting SNI/ALPN values and local proxy tunneling happen transparently in most
+cases e.g. when using `tsh` client to connect. For some clients (for example,
+native/GUI database clients like `psql` or pgAdmin instead of `tsh db connect`)
+the local proxy needs to be running so these clients can connect to it.
+
+### Diagram
+
+![TLS routing](../../img/architecture/tls-routing.png)
+
+Let's take a look at how each particular protocol Teleport supports implements
+TLS routing.
+
+## SSH
+
+When connecting to an SSH node using `tsh ssh` command, it first dials Teleport
+proxy over TLS and requests `teleport-ssh-proxy` ALPN protocol.
+
+No local proxy is started in this case as `tsh` uses this TLS connection as a
+transport to establish the SSH connection.
+
+### OpenSSH
+
+To support standard OpenSSH client, Teleport provides a `tsh proxy ssh` command
+which can be used as a `ProxyCommand`.
+
+Similarly to `tsh ssh`, `tsh proxy ssh` establishes a TLS tunnel to Teleport
+proxy with `teleport-ssh-proxy` ALPN protocol, which `ssh` then connects over.
+
+See [OpenSSH client](../server-access/guides/openssh.mdx#use-openssh-client)
+guide for details on how it's configured.
+
+## Reverse tunnels
+
+Reverse tunnel agents such as node, application and database agents as well as
+proxies of leaf clusters work in similar fashion to SSH.
+
+An agent first opens a TLS tunnel to the cluster's proxy with
```

Comment:
Reverse tunnel agents for Teleport node, app and database services open a TLS ...